From ea75bd4122c172e2762d9ea2e70b507174b5140f Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Thu, 28 Apr 2022 19:08:50 +0300
Subject: [PATCH] [SW-82716] common: add sync support for i2c

When SRAM_RUN is replied with status OK, arcboot is not syncing
with the host/BMC that it got the status and thus procedes to executing
the image.

To fix the timeout - add special sync code which is invoked on RUN_OK
status and delays the NO_RETURN execution by 200ms for the BMC to read
the status value.
After the timer expiration, the NO_RETURN command is executed and we
load the provided FIT which should take over the communication.
200ms was chosen as a relatively large and safe value.
Since this is not a time short critical execution path, we can use that
large value.

Change-Id: I17fb333e6b502390d599f1297363620c9fe20f7e
---

diff --git a/common/module_fwupd.c b/common/module_fwupd.c
index fc02b2a..4dc3d18 100644
--- a/common/module_fwupd.c
+++ b/common/module_fwupd.c
@@ -16,6 +16,7 @@
 #include <fit_load_type.h>
 #include <security.h>
 #include <memory.h>
+#include <delay.h>
 
 #include "cmd_module.h"
 #include "fit.h"
@@ -61,6 +62,18 @@
 static struct fwupd_state i2c_state;
 static struct fwupd_resp_prop resp_props[RESP_PROP_NUM];
 
+#define FWUPD_NORET_DELAY_US	200000
+
+static void fwupd_sync_exec(struct fwupd_state *state)
+{
+	if (state->sync_exec_res == NOT_FOUND || !state->noret_delay)
+		return;
+
+	if (arcboot_rom_timeout_expired(state->noret_delay, FWUPD_NORET_DELAY_US)) {
+		state->exec_res = state->sync_exec_res;
+		state->sync_exec_res = NOT_FOUND;
+	}
+}
 
 uint32_t i2c_fwupd_cmd_readl(void)
 {
@@ -332,6 +345,7 @@
 	if (state->sending_error_flag && arcboot_error_is_pending())
 		return FOUND_SLOW_RETURN;
 
+	fwupd_sync_exec(state);
 	if (state->exec_res != NOT_FOUND)
 		return state->exec_res;
 
@@ -520,7 +534,13 @@
 	if (err)
 		return err;
 
-	state->exec_res = FOUND_NO_RETURN;
+	if (state->synced_no_return) {
+		state->sync_exec_res = FOUND_NO_RETURN;
+		state->noret_delay = arcboot_rom_get_ts_us();
+	} else {
+		state->exec_res = FOUND_NO_RETURN;
+	}
+
 	return err;
 }
 
